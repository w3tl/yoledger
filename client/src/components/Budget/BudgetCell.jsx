import React from 'react';
import PropTypes from 'prop-types';
import AmountInput from '../AmountInput';

class BudgetCell extends React.Component {
  static propTypes = {
    amount: PropTypes.number,
    onUpdate: PropTypes.func,
  }

  constructor(props) {
    super(props);
    this.state = {
      amount: props.amount,
    };
  }

  handleChange = ({ target }) => {
    this.setState({ amount: target.value });
  }

  handleSave = () => { // TODO: save only when changed
    const { amount } = this.state;
    if (amount) {
      const { onUpdate } = this.props;
      onUpdate({ amount });
    }
  }

  render() {
    const { amount } = this.state;
    return (
      <div style={{ height: 30 }}>
        <AmountInput value={amount} onChange={this.handleChange} onBlur={this.handleSave} />
      </div>
    );
  }
}

BudgetCell.defaultProps = {
  amount: null,
  onUpdate: () => {},
};

export default BudgetCell;
