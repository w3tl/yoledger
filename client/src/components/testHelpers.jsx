import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
