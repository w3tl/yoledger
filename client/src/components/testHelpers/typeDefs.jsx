export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]

  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
`;
