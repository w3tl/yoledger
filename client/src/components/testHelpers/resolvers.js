import dateResolver from '../../../../graphql/date';
import assets from '../Asset/mockData';
import { accounts, periods, budgets } from '../Budget/mockData';

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      return assets.find(a => a.id === id);
    },
    accounts(root, { type }) {
      if (type === 'ASSET') return assets;
      return [];
    },
    budgets() {
      return { accounts, periods: periods.map(dateStr => new Date(dateStr)) };
    },
    budget() {
      return budgets[0];
    },
  },
  Mutation: {
    upsertBudget() {
      return {
        success: true,
        allocation: {
          account: { id: '99', name: 'new account' },
        },
      };
    },
  },
};
