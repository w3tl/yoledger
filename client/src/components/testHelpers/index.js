import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
