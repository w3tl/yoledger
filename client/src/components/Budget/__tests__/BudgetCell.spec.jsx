import React from 'react';
import { mount } from 'enzyme';
import BudgetCell from '../BudgetCell';
import { testRenderWithoutError } from '../../testHelpers';

describe('BudgetCell', () => {
  testRenderWithoutError(<BudgetCell amount={10} />);

  it('should handle the save button click', () => {
    const onUpdate = jest.fn();
    const output = mount(
      <BudgetCell amount={10} onUpdate={onUpdate} />,
    );

    output.find('input').simulate('blur');
    expect(onUpdate).toHaveBeenCalledWith({ amount: 10 });
  });
});
