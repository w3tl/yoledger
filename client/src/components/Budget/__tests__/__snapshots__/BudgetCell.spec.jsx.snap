// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetCell renders without error 1`] = `
<div
  style={
    Object {
      "height": 30,
    }
  }
>
  <Component
    onBlur={[Function]}
    onChange={[Function]}
    value={10}
  />
</div>
`;
